### DISCUSSION TITLE
Best way to write entities from a single uow perspective

### DISCUSSION THREAD

**Discussion by Gomez12** [ravendb/ravendb#19050] [Q&A]
We have a message bus which does steps, and now we have the wish to save some uow states so we can in a later process retrieve them.

We are using the rebus library to be bus-agnostic, but rebus delivers messages one by one, over multiple threads.

What is the recommended way to save entities to raven in such a scenario?
We have tried sessions, store 1 entity and then savechanges, but then the message throughput goes from peak 2000/s to something like peak 400/s and currently we are testing it in dev with a embedded RavenDB, so I would guess no network latency and no security.

We have tried bulk insert with single entities, but it doesn't help.

We don't want to use an in-memory queue to collect multiple entities and then periodically bulk insert them, because of the change of losing entities which have been processed by power outages etc.

Are we missing a method here? We only want to save off 1 item at a time.

Or should we be searching for a disk-backed queue?

**Comment by ayende** [comment]
If you are using concurrent threads to process this, then the way to go about this is to do call merging.
Basically, any thread that completed a step will persist its state (along with any other) in one shot.

Here is an example on how to do that:

https://ayende.com/blog/176257/10x-speedup-utilizing-nagle-algorithm-in-business-application
